### ISSUE TITLE
Universe Generation Crash / Bad Version String when using Slackware 14.1 Custom Build Script

### ISSUE THREAD

**Issue by ghost** [freeorion/freeorion#323]
freeorion 0.4.5-RC1

The game doesn't work.

I start it by typing `freeorion` in command line. It launches, gets to the main menu. I select "Quick Start". After that the Messages window prints a "Creating AI Clients" message, then a "Generating Universe" message and after that an error window saying "The connection to the server has been lost".

[Sreenshot](http://postimg.org/image/qotnx5lb9/)

**Comment by geoffthemedio** [comment]
What OS are you using? The version string / application title are oddly broken.

Please clear the contents of the directory described here:

http://freeorion.org/index.php/Config.xml

Then attempt to start a game with 1 AI and otherwise default settings.

After it crashes, go to the same directory and locate all the .log files, and post them somewhere (eg. pastebin) and link to them in a reply.

**Comment by ghost** [comment]
I using Slackware 14.1 x86_64 'stable'

Delete the folder ~/.freeorion and i restart the game again and the problem is the same.

These are the logs:
[AI_X logs](http://pastebin.com/09LhVFQk)
[Config.xml](http://pastebin.com/Abkinc2F)
[freeorion.log](http://pastebin.com/VAU7hG0E)
[freeoriond.log](http://pastebin.com/aXK2vGrw)

**Comment by geoffthemedio** [comment]
How did you build FreeOrion? Given the broken version string in the screenshot, you might not have run CMake properly.

Aside from that, much of the galaxy setup seems to have run OK, but it ends suddenly after adding starting buildings.

**Comment by ghost** [comment]
Maintain this game for Slackware.

This is the [build script](http://slackbuilds.org/slackbuilds/14.1/games/freeorion/freeorion.SlackBuild)

**Comment by geoffthemedio** [comment]
How are you running FreeOrion? From what directory?

The crash appears to be happening when the Python scripts setting up the player homeworlds tries to load "starting_buildings.txt" from the resource directory. Given that you have weird errors in the version string, I'm guessing you haven't built and installed FreeOrion as it is expecting, so it can't find that file.

**Comment by geoffthemedio** [comment]
Is that build script running CMake? If not, you might need to modify it to do everything that the CMake script is doing, including putting content files where they're expected.

Alternatively, adding the resource directory to the path so the Python scripts can find "starting_buildings.txt" and similar files it might help. Apparently the in-Python path resolution isn't working the same as the in C++ equivalents, for you.

**Comment by ghost** [comment]
A solution would you suggest ?

**Comment by geoffthemedio** [comment]
Use CMake to build and install instead of whatever that Slackware-specific script is doing.

Or try adding the resources directory to the system path. That might be the easiest / simplest, though might just reveal another related problem given the odd installation.

Or wait for someone else who knows more about the CMake build system and how it works on Linux or Slackware in particular to comment.

That all said, there should be some error reporting if that string list load fails, and I don't know why you wouldn't see that in the server log file, so there might be something else going on.

**Comment by Vezzra** [comment]
Ok, a few things:
- @dslackw: Don't use the source tarball github creates automatically. For Linux distros we provide a manually created source tarball which contains an already correctly generated version.cpp. The one for RC1 is [FreeOrion_v0.4.5-RC1_2015-08-24.5fc20ff_Source.tar.gz](https://github.com/freeorion/freeorion/releases/download/v0.4.5-rc1/FreeOrion_v0.4.5-RC1_2015-08-24.5fc20ff_Source.tar.gz) (you may want to wait a few hours, today evening I will produce RC2, I suggest trying again with that). When you use these source tarballs, you can remove the lines from the build script which copy util/version.cpp.in to util/version.cpp (or better: you must, otherwise the pre-generated version.cpp is overwritten with the template file!).
- As far as I can glance from the build script, it definitely calls cmake, so the freeorion build scripts _should_ be executed correctly.
- I doubt that the issue here is that the Python script can't find and open this file. That should produce a respective error message in the log, not a crash like this. The problem is, we can't be sure that this operation is actually the one which causes the crash, because in the past there have been cases where some portion of the log file had been lost in case of such a crash. I know we use another logging library now, but still, I don't trust these things entirely. One way to check that is to try to run FO several times and examine freeoriond.log afterwards. If the last line in the log is always the same, that would indicate that indeed this is the operation where the crash occurs. If it varies, well, we've got a problem...

Well, whatever, first thing I'd suggest doing now is trying again with RC2 which is going to be released today, and this time using the manually created source tarball. Second, if FO still crashes (which is probably to be expected), launch it several times, check freeoriond.log if the last lines are always the same or if they vary, and report back.

**Comment by ghost** [comment]
Ok... Thank you,
I will wait for the RC2 released and try again to build.

**Comment by geoffthemedio** [comment]
Boost::Log is being set to auto_flush = true which should result in logs being written immediately. ("should")

Another thing to try is adding
'print "Loaded and added buildings"

to line 427 of empire.py to see if it crashes before or after finishing adding the buildings.
https://github.com/freeorion/freeorion/blob/master/default/python/universe_generation/empires.py#L427

**Comment by Vezzra** [comment]
Hm, I think it would make more sense to add

```
print "...adding building", building
```

between line 425 and 426, as part of the loop, like this:

```
    for building in load_string_list(os.path.join(fo.get_resource_dir(), "starting_buildings.txt")):
        print "...adding building", building
        fo.create_building(building, homeworld, empire)
```

because I don't think adding a print statement _after_ the loop has finished will help much - after all, there is a print statement on the next line that apparently didn't get executed, which already indicates that the loop never finished.

Adding the print statement within the loop will give us some more precise info on where exactly the crash happens - if the print statement never executes, then it's the `load_string_list` call that fails, otherwise it's the `fo.create_building` call, and we'll even get to know at which building this call fails.

**Comment by Vezzra** [comment]
RC2 is up now.

**Comment by ghost** [comment]
Try to build without rename  `mv util/Version.cpp.in util/Version.cpp`

http://pastebin.com/gKRsDhRg

**Comment by ghost** [comment]
Now i try to build from sourceforge tarball https://sourceforge.net/projects/freeorion/files/FreeOrion/FreeOrion%20Version%200.4.5/

**Comment by ghost** [comment]
Build normal from sourceforge tarball but the error is the same "The connection to the server has been lost".
[Screenshot](http://postimg.org/image/6rlwclpen/)

**Comment by geoffthemedio** [comment]
Wouldn't have expected otherwise. See above suggestions to try adding logging to the building adding part of the Python universe generation.

**Comment by ghost** [comment]
[freeoriond.log](http://pastebin.com/VTnnekht)

**Comment by geoffthemedio** [comment]
What did you change to produce the new log?

**Comment by ghost** [comment]
After building the release version RC2 deleted directory ~ / .freeorion and ran the FO around 7 times.

**Comment by ghost** [comment]
This is latest FO configuration http://pastebin.com/5wpS7np8

**Comment by geoffthemedio** [comment]
Please try adding the logging line suggested by Vezzra above, to narrow down the location of the crash.

**Comment by Vezzra** [comment]
@dslackw, I produced a special source tarball with the additional logging commands, you can download it [here](http://we.tl/0wONJKtUl5) (please note: this link only works for a week).

Try again with that source tarball and post the contents of freeoriond.log after the crash.

**Comment by ghost** [comment]
Thank you, i will try it now...

**Comment by ghost** [comment]
[freeorion.log](http://pastebin.com/BXMfQWtW)
[freeoriond.log](http://pastebin.com/h8bpRmjC)

**Comment by Vezzra** [comment]
Apparently you didn't use the source tarball I linked to (the server log contains error messages that are caused by a bug that has been fixed in the version of this source tarball), and also didn't remove the lines from your build script that copy util/version.cpp.in to util/version.cpp.

Please, try again and use this source tarball: [FreeOrion_2015-09-01.e3ab1c5_Source.tar.gz](http://we.tl/0wONJKtUl5). This is a special tarball with additional logging commands, which are not contained in the tarballs you can download from github or sourceforge. _You have to use the link I provided._

And please also remove these lines from your build script:

```
# Fix cmake target for version file
mv util/Version.cpp.in util/Version.cpp
```

Otherwise the version string isn't correctly set, and we can't verify if the log files you provided really have been generated by the correct build.

**Comment by ghost** [comment]
Use, maybe something went wrong... no problem i try again...

**Comment by ghost** [comment]
I removed folder /build, source code and package, download again and build so: 

[freeorion.log](http://pastebin.com/mkmGxb2G)
[freeoriond.log](http://pastebin.com/BfVcULWD)

**Comment by Vezzra** [comment]
Ok, much better :smiley: 

According to that log, the crash does not occur when trying to access the starting_buildings.txt file, but when trying to add the first building. That means it's the `fo.create_building` function that fails.

I'll add some more logging commands and provide a link to another special source tarball.

**Comment by Vezzra** [comment]
New source tarball: [FreeOrion_2015-09-02.6cdbeb3_Source.tar.gz](http://we.tl/mVr3lVTRNc)

Please try again with this new tarball, same procedure. Link will be valid for a week.

**Comment by ghost** [comment]
Ok... Done

[freeorion.log](http://pastebin.com/kWbCjFHg)
[freeoriond.log](http://pastebin.com/BGqV4eRe)

**Comment by Vezzra** [comment]
Hm, according to the log it's the actual call to `Universe::CreateBuilding` that crashes. Trying to track down exactly where things go wrong is going to be tedious with this method. Is there any possibility to get a backtrace of the crashed freeoriond process on your machine? That might save us a lot of work...

**Comment by ghost** [comment]
I am willing to help this very good game for Slackware users. What can I do;

**Comment by Vezzra** [comment]
> What can I do

Well, a backtrace of the crash would be helpful, alas, I'm not on Linux, so I don't know how you can get one on your system for a crashed process. Maybe some of our Lunix folks can help?

Anyway, unless we can figure out how to get that backtrace, we have to continue with the special source tarballs. Here is the next one, with some more additional logging:

[FreeOrion_2015-09-07.c79159e_Source.tar.gz](http://we.tl/KR6vDNWi7y)

Again, link only valid for about a week. Proceed as before and post freeoriond.log after crash with the new source tarball.

**Comment by ghost** [comment]
[freeoriond.log](http://pastebin.com/BKTdDGsH)

**Comment by ghost** [comment]
Can you describe that is the problem? Lost files on the building? maybe I can create a patch.

**Comment by pitchforks** [comment]
@Vezzra the problem with getting a backtrace is that it's not freeorion that crashes. So one can't just run freeorion with gdb, wait for it to crash then request the backtrace. It's freeoriond that crashes. For somebody who isn't very familiar with gdb and not familiar with C++ and particularly freeorion codebase at all - it's even hard to figure out the exact way that freeoriond is started by freeorion behind the curtain. Is it a thread? A fork? I have tried to take guesses and add different kinds of catchpoints to gdb with not much luck. The problem with getting a backtrace of freeoriond is that it crashes very quick, so there's no time to find its PID using ps and attach gdb to it. Perhaps, in addition to this custom source code tarball, you can also provide a small patch that adds a hardcoded delay at the very beginning of freeoriond? So one can apply it and gain the time required for the investigation mentioned above before freeoriond crashes?

**Comment by geoffthemedio** [comment]
It's possible to start freeoriond separately, then have a freeorion process connect to it. run

`freeoriond`

and then separately run

`freeorion --force-external-server`

and quickstart.

**Comment by pitchforks** [comment]
@geoffthemedio thank you, this makes it possible to attach directly to freeoriond

I have managed to get a backtrace

http://www73.zippyshare.com/v/91AoOmu6/file.html

I did build the latest special source tarball with debug symbols by using

```
-DCMAKE_BUILD_TYPE=Debug ..
```

and not stripping anything.

However, I had to strip the freeoriond binary manually after installing it, otherwise gdb would eat all my RAM when trying to load it. Also, the attached backtrace is the result of `backtrace` gdb command, because `backtrace full` kills my OS by eating up all the RAM as well.

HTH

**Comment by geoffthemedio** [comment]
It looks to be a crash somewhere in the content parsing code.

What version of Boost are you building against?

**Comment by pitchforks** [comment]
Slackware 14.1 has boost-1.54.0

**Comment by geoffthemedio** [comment]
You might need to update to Boost 1.56, or better yet 1.58. The main motivation for wanting 1.56 was related to logging, which you don't appear to have problems with, but it might help.

If that doesn't help, you might have a compiler issue with the parser code... in that case we probably can't do much about it.

**Comment by Vezzra** [comment]
Ok, let me recap what I think is happening here: When trying to create the first building, the BuildingTypeManager gets initialized, and because of that building.txt gets parsed. Some weird thing happens deep in the parser code and the freeoriond process crashes with a segfault.

That makes no sense at all. That code is part of the common library, so this exact piece of code has been already executed by the human client, freeorion, at that point. IIRC the human client reads in and parses the content files right on startup, before any game (and thereby the server) can be launched.

Why would the same piece of code execute perfectly fine in the client process, and produce a weird segfault crash in the server process? This is very puzzling...

**Comment by geoffthemedio** [comment]
Content files are read and and parsed when their respective content type is accessed, not immediately at startup. For buildings in particular, there's probably nothing that would cause them to be parsed by the client before receiving a turn update.

**Comment by Vezzra** [comment]
Hm, ok, in that case things aren't as weird as I thought. Which is good.

@dslackw, @pitchforks: Has one of you tried Geoffs suggestion of building FO against a higher boost version? Any success with that?

**Comment by pitchforks** [comment]
@Vezzra first of all, please edit the issue title and add the Slackware version to it. Make it "Slackware 14.1" instead of just "Slackware". This will make it easier to find it out in the future, if it will be needed for reference.

As for trying to build against a higher boost version - here is my perspective: I am very reluctant to upgrade boost, since it comes with stock Slackware and I don't want to end with a newer version of boost eventually breaking those packages that depend on it. However, the -current branch of Slackware - the one that will become the next stable Slackware version when it will be released (there's no release date fixed yet) - seems to have boost 1.58. Unfortunately, I lack a second computer where I could install -current and try to compile freeorion. I also don't have sufficient hardware resources to try doing this in a virtual machine. So unfortunately this is not an option to me. We'll see what @dslackw has to say, since he's the maintainer of freeorion and his opinion has more weight. I've just stepped in trying to give a bit of additional help.

And what about https://github.com/freeorion/freeorion/commit/1989c61ab588a0e65509855aa8da0122f6bd2489 ? Is it related to this bug?

Also, I'm wondering if you know of anybody else who managed to build a functional freeorion version using this exact version of boost. Was a newer version of boost suggested to see if that may help or because it is known for sure that this version of boost won't work after all this long digging process? I would not mind if you decide to make boost 1.56 or 1.58 a mandatory build dependency and enforce it by adjusting cmake's config file. This would be fair if you think it's appropriate and it will mean that current stable Slackware version can't provide the required dependency and we'll have to wait a bit till next is out.

To summarize: I can only keep trying builds with this version of boost if you have any hope you can make it work.

**Comment by Vezzra** [comment]
> please edit the issue title and add the Slackware version to it. Make it "Slackware 14.1" instead of just "Slackware".

Done.

> And what about 1989c61 ? Is it related to this bug?

Highly unlikely. The code section in question tried to deal with a theoretical edge case that most likely never ever has actually occurred, so the buggy code never had been executed. And even if, hitting the 2 billion objects mark won't happen during universe generation of a quickstart game... :wink: 

> Also, I'm wondering if you know of anybody else who managed to build a functional freeorion version using this exact version of boost.

Apparently, our 0.4.5 release [has been packaged for openSUSE](http://www.freeorion.org/forum/viewtopic.php?f=24&t=9757) and been built on that distro against boost 1.54 without any significant issues. So it seems unlikely that this is even the problem here, but to be sure to rule that out we nevertheless have to try it.

As to the challenges you mention with regard to building against an upgraded version of boost, I can certainly understand your reluctance. Unfortunately I can't assist much further in resolving the issue in this case, as I'm not on Linux at all, and my proficiency in Linux is severely limited.

Funny thing is, I actually _have_ the resources at my hand to set up a VM with Slackware - I've access to a Hyper-V server with which that should be possible. So, you've the knowledge, but not the resources, and I have the resources, but lack the knowledge... life can be cruel sometimes :wink:

**Comment by ghost** [comment]
I do not think is boost issues. Before ten days built the FO in Slackware 14.1 -current and I got the same problem.

**Comment by pitchforks** [comment]
@dslackw perhaps you can try getting a backtrace of freeoriond from -current and boost 1.58, maybe comparing it with the one I've got for 14.1 and boost 1.54 can provide additional insights for freeorion developers?

**Comment by geoffthemedio** [comment]
What compiler are you using?

**Comment by pitchforks** [comment]
Slackware 14.1 has gcc-4.8.2

**Comment by geoffthemedio** [comment]
What happens if you delete the contents, but leave an empty building.txt ?

**Comment by pitchforks** [comment]
@geoffthemedio please elaborate. There's a `default/buildings.txt` file in the source code tarball. It seems to go to `/usr/share/freeorion/default/buildings.txt` in the resulting package. Do you suggest to remove all the content from that file, leaving it empty, run freeorion, start a quick game and see what happens (if freeoriond crashes as well)?

**Comment by Vezzra** [comment]
Exactly, and if it crashes, post the freeoriond.log and the backtrace.

**Comment by ghost** [comment]
This is [configuration](http://pastebin.com/pTu6yGEU) from Slackware 14.1 -current x86 with boost 1.58.0

Compiler version:

```
slack@darkstar ~> gcc --version
gcc (GCC) 4.9.3
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

And [freeoriond.log](http://pastebin.com/66DBajHd)

**Comment by pitchforks** [comment]
This is Slackware 14.1 and the same package I've used to produce the backtrace above. The only difference is that I've emptied `/usr/share/freeorion/default/buildings.txt` as suggested above.

freeoriond did not crash this time after quickstart!

And here are the logs from `~/.freeorion`:

[freeorion.log](http://pastebin.ca/3157142)

[freeoriond.log](http://pastebin.ca/3157145)

**Comment by geoffthemedio** [comment]
Try again with just the following in buildings.txt

```
BuildingType
    name = "BLD_GENOME_BANK"
    description = "BLD_GENOME_BANK_DESC"
    buildcost = 15
    buildtime = 16
    location = And [
        Planet
        Not Contains Building name = "BLD_GENOME_BANK"
        OwnedBy empire = Source.Owner
        TargetPopulation low = 1
    ]
    EnqueueLocation = [[ENQUEUE_BUILD_ONE_PER_PLANET]]
    icon = "icons/building/genome_bank.png"
```

**Comment by ghost** [comment]
@geoffthemedio Yes this is work for me.

**Comment by geoffthemedio** [comment]
Try

```
BuildingType
    name = "BLD_SUPER_TEST"
    description = "BLD_SUPER_TEST_DESC"
    buildcost = 1
    buildtime = 1
    location = Planet
    effectsgroups = [
        Effectsgroup
        scope = And [
            Object id = Source.PlanetID
            Planet
        ]
        effects = [
            SetSpecies name = "SP_SUPER_TEST"
            SetPopulation value = max(Target.Population, 1)
        ]

        EffectsGroup            // remove building when done
            scope = Source
            activation = Species name = "SP_SUPER_TEST"
            effects = Destroy
    ]
    icon = "icons/species/other-04.png"


#include "shared_macros.txt"
#include "col_buildings.txt"
```

**Comment by ghost** [comment]
And this is [freeoriond.log](http://pastebin.com/5RLyhJAB) from Slackware 14.1 stable x86_64

**Comment by ghost** [comment]
From second example [freeoriond.log](http://pastebin.com/xQifnwLY)

**Comment by geoffthemedio** [comment]
Try with this buildings.txt

http://pastebin.com/9s9Dk8TU

**Comment by ghost** [comment]
[screenshot](http://tinypic.com/view.php?pic=20iyicm&s=8#.VfETI3WlxBc)
[freeoriond.log](http://pastebin.com/t8NQ8t1R)

**Comment by geoffthemedio** [comment]
That was the first two buildings, the comments at the start, and the #includes at the end. Try adding more buildings (or deleting less and less) until you find out what makes it crash... might be a particular building, or a total length of the file that causes the crashing.

**Comment by ghost** [comment]
I go to sleep now and will try later, Thank you.

**Comment by pitchforks** [comment]
A very preliminary result: the problem doesn't seem to be in buildings.txt, but rather in a file it includes: shared_macros.txt

freeorion stops crashing when I manually empty this included file.

Hopefully I can dig deeper since this file seems to have small size. I will report back if I manage to find anything.

**Comment by pitchforks** [comment]
Here are 3 different versions of file `shared_macros.txt`:

[version 1](https://gist.github.com/pitchforks/da99b9097dc6d8053334)
[version 2](https://gist.github.com/pitchforks/3f8675983be9dca40d96)
[version 3](https://gist.github.com/pitchforks/e0cdc7a0c4c52c5a99ef)

version 1 is a slightly trimmed change of vanilla file from the source code tarball - just a few comments removed

version 2 + version 3 = version 1, version 2 being the long head and version 3 being the short tail of version 1

freeorion crashes with version 1

freeorion does not crash with version 2

freeorion does not crash with version 3

so it looks like two parts of this file don't play nice together

any further suggestions?

**Comment by geoffthemedio** [comment]
What happens if you remove the line at the end,
`#include "col_buildings.txt"`

**Comment by pitchforks** [comment]
@geoffthemedio the presence or absence of this line in `buildings.txt` doesn't matter - freeoriond crashes in both cases - when the include is present in the vanilla version, or when I delete that include line manually.

Like I said above, the troublemaker is the file `shared_macros.txt`

Just for the record - after each manual change of said text files - I have fully restarted freeorion, just to exclude any caching getting involved.

**Comment by geoffthemedio** [comment]
What if you put the ENQUEUE_BUILD_ONE_PER_PLANET part before the DESCRIPTION_EFFECTSGROUP_MACRO part?

**Comment by pitchforks** [comment]
@geoffthemedio

it still fails if I do that

**Comment by geoffthemedio** [comment]
try removing the line
`description = "@1@"`
?

**Comment by pitchforks** [comment]
@geoffthemedio vanilla `shared_macros.txt` without this line still crashes. But as said above, freeorion does not crash with [this](https://gist.github.com/pitchforks/3f8675983be9dca40d96) version (it's version 2 above) of the file. And this version has this line.

**Comment by geoffthemedio** [comment]
I know what you wrote above, but it doesn't make sense that this macro definition would cause problems just with parsing buildings.txt and only with the following macro also present. Particularly so since the DESCRIPTION_EFFECTSGROUP_MACRO macro isn't even used in buildings.txt or any of the files it includes. More so when nobody else has reported this problem except when also building/running on Slackware.

So, I'm trying to figure out if there's something else going on, like the fully-included buildings.txt being too long, or something unusual about the identified macros being a source of the problem, like the @1@ line. 

I would again/still like to know what happens with a newer version of Boost.

**Comment by geoffthemedio** [comment]
Any effect from removing the comment starting with `// This Effect is` ?

**Comment by pitchforks** [comment]
@geoffthemedio 

> like the fully-included buildings.txt being too long

Do you mean the content from `buildings.txt` itself and the content from the files it includes at the bottom? If so, I can see your point. Though, since I don't understand the inner works of the parsing process - shouldn't a too big content cause out of memory errors rather than segfaults? I'm not noticing any excessive memory usage when running freeorion the regular way. I did notice excessive memory usage when I was running freeoriond with gdb, but I'm not sure that's related - those happened because of debug symbols enabled in all freeorion code were generating lots of additional backtrace details I suppose?

> I would again/still like to know what happens with a newer version of Boost.

Perhaps @dslackw can repeat all these tests in his Slackware -current instance, I can only do the tests for 14.1 and boost 1.54

> Any effect from removing the comment starting with // This Effect is ?

vanilla `shared_macros.txt` with just this comment removed still produces a negative result

**Comment by Vezzra** [comment]
Sorry for reporting back only now, but I've been trying to install a Slackware 14.1 system in a VM and build and run FO there. Me being the Linux noob I am, that took some trial and error, but right now FO is compiling (hopefully it will complete successfully, that remains to be seen). @dslackw, @pitchforks, assuming I'll get the same crashes as you, I might need some assistance/instructions at installing a newer boost version.

**Comment by Vezzra** [comment]
Good and bad news. Good news: FO built successfully. Bad news: when I launch FO, I get the following error messages on the command line:

```
ALSA lib confmisc.c:768:(parse_card) cannot find card '0'
ALSA lib conf.c:4248:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4248:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1251:(snd_func_refer) error evaluating name
ALSA lib conf.c:4248:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:4727:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2239:(snd_pcm_open_noupdate) Unknown PCM default
AL lib: (EE) ALCplaybackAlsa_open: Could not open playback device 'default': No such file or directory
Segmentation fault
```

I guess that might have to do with running inside a VM and not having sound? How can I work around that?

**Comment by geoffthemedio** [comment]
try blanking buildings.txt; perhaps the segfault is the topic of this issue, and the AL lib issues are unrelated?

**Comment by Vezzra** [comment]
Oh, sorry, these error messages have nothing to do with the issue itself. They occur immediately when I try to launch the human client and are printed to the console, freeorion.log contains only two lines. I've been posting them here in hope that @dslackw or @pitchforks might have any ideas what's going wrong here and how I can work around it (as it looks like something doesn't work right here with the alsa lib installation on Slackware 14.1).

**Comment by pitchforks** [comment]
@Vezzra 

Are you running freeorion as regular user that you eventually did not add to the `audio` group?

**Comment by Vezzra** [comment]
@pitchforks: yes and yes, I did not add that user to the audio group. Doing this didn't help though, I still get the same error messages.

However, I think @geoffthemedio might be right after all, at least insofar as the segfault might not be related to the alsa lib issues. Unfortunately, as I already said, the failure occurs very early during program startup, freeorion.log only contains two lines which give no hint at all at the cause of the problem.

Which means I'm most likely screwed, as my know how is far to limited to tackle this kind of issue on a Linux system. I can only tell what I did to get FO built, maybe I did something wrong.

The main problem for me was to get the dependencies on SDL2 and OpenAL installed. `slackpkg` didn't get me far, because its search function didn't yield any packages for SDL2 or OpenAL. So I went to pkgs.org, where I found .txz packages for both libraries. Installing them with `upgradepkg --install-new` went absolutely smoothly, after that the `cmake && make && make install` procedure went without any hassles. Note: I ran cmake with the same options it's invoked with in the build script @dslackw posted above.

Running the resulting `freeorion` gives me the error messages I posted above.

**Comment by pitchforks** [comment]
@Vezzra diving into an unknown Linux distro without much prior experience is probably as challenging as trying to build freeorion without knowing the codebase and generally much programming - and hitting a nasty bug like this one :)

> So I went to pkgs.org

This is certainly not the best approach, but I won't comment on it to avoid derailing the topic a lot.

Perhaps we can find a way to deal with these long delays between posts here by chatting interactively? I was pointed to a forum thread where you say you don't do IRC, it can be Jabber or tox then. If you're willing to push this further, perhaps we can catch eachother online somehow?

I'm not thinking about any arranged meeting at certain hours, just that it would be nice if we can talk interactively when we happen to be both online.

**Comment by Vezzra** [comment]
> diving into an unknown Linux distro without much prior experience is probably as challenging as trying to build freeorion without knowing the codebase and generally much programming - and hitting a nasty bug like this one

And not to forget, only having so much time I can allocate to this problem... :wink:

> > So I went to pkgs.org
> 
> This is certainly not the best approach, but I won't comment on it to avoid derailing the topic a lot.

Yeah, I've already been wondering how far I should continue discussing this side-issue here. You're probably right, we should continue this elsewhere. I can open a thread on [our forum](http://www.freeorion.org/forum/), you'll need to register there.

> Perhaps we can find a way to deal with these long delays between posts here by chatting interactively? I was pointed to a forum thread where you say you don't do IRC, it can be Jabber or tox then.

Hm, the problem is, the only IM program I'm actually using is Skype, and even there I'm only online when I actually need it. Furthermore, for purposes like these I only use the account of my cyber-alter-ego Vezzra (I try to avoid revealing my real identity), with which I'm almost never online - pratically only when making specific appointments (which _does_ happen, I already had chat sessions for FO purposes).

But we should probalbly continue this discussion too elsewhere. As github lacks PMs, either per email (a public email should be visible at my profile), or on our forum via PM?

**Comment by Vezzra** [comment]
Ok, I've found a fix for the issue. @geoffthemedio was spot on:

> you might have a compiler issue with the parser code

That's apparently exactly what's going on here, gcc somehow messes something up in the parser code.

Fortunately, Slackware 14.1 also ships with clang/llvm, and when you build FO with that toolchain everything works fine. The question now is, if it's mandatory for the build script to use gcc, or if switching to clang/llvm is ok. The following adjustments to the build script are required:

1) Add the following lines somewhere at the beginning of the script, before the invocation of cmake:

```
export CC=/user/bin/clang
export CXX=/usr/bin/clang++
```

CMake honors those environment variables and consequently uses clang instead of gcc.

2) Add the following option to the invocation oc cmake: `-D_CMAKE_TOOLCHAIN_PREFIX=llvm-`, so the cmake invocation looks like this (important: please note the underscore between the '-D' and 'CMAKE' parts!):

```
  cmake \
    -D_CMAKE_TOOLCHAIN_PREFIX=llvm- \
    -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
    -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
    -DCMAKE_INSTALL_PREFIX=/usr \
    -DLIB_SUFFIX=${LIBDIRSUFFIX} \
    -DCMAKE_BUILD_TYPE=Release ..
```

This apparently is required to tell cmake to use the llvm toolchain. Building FO with this modifications to the build script produced a working build.

@dslackw, @pitchforks, can you test on your systems if this fix works for you?

If you can't use clang/llvm for building packages, we're out of luck, unfortunately. Fixing gcc issues is beyond our possibilities, sorry.

**Comment by geoffthemedio** [comment]
I would once again suggest trying a newer version of Boost. It's possible they noticed the issue and fixed it in a later release.

**Comment by ghost** [comment]
Thanks everyone, for your patience and for the great job, works fine now.

@geoffthemedio at present I can not use the latest version of boost is not available at release Slackware 14.1 -stable.

So i have update freeorion with version 0.4.5 in the repository SlackBuilds.org and it is available next week after public update.

[freeorion.SlackBuild](http://slackbuilds.org/cgit/slackbuilds/commit/?h=user/dslackw/updates)

Best regards,
Dimitris

**Comment by geoffthemedio** [comment]
Are you unable to download and build Boost from source?

**Comment by Vezzra** [comment]
According to what @dslackw said earlier:

> I do not think is boost issues. Before ten days built the FO in Slackware 14.1 -current and I got the same problem.

And @pitchforks mentioned:

> However, the -current branch of Slackware - the one that will become the next stable Slackware version when it will be released (there's no release date fixed yet) - seems to have boost 1.58.

This would indicate that @dslackw already tried to build against boost 1.58, which didn't solve the issue. So it really looks like we're dealing with a compiler issue here (I mean, building with clang/llvm works after all...).

Anyway, we have a working solution, so I'm closing this issue.

**Comment by MagaTailor** [comment]
The solution probably works but the slackbuild author must have messed the clang/llvm architecture flags and the alleged i586 build crashes with SIGILL on my non-SSE2 machine.

I contacted the author of the script about it and he mentioned this issue here. If there's a standard way of passing the clang march flags I'm sure he'll be adding it to his script.
Thx.

**Comment by Vezzra** [comment]
Hm, the only thing that caught my attention when looking at the `freeorion.SlackBuild` script are the following lines (45-50):

```
if [ "$ARCH" = "i486" ]; then
  SLKCFLAGS="-O2 -march=i486 -mtune=i686"
  LIBDIRSUFFIX=""
elif [ "$ARCH" = "i686" ]; then
  SLKCFLAGS="-O2 -march=i686 -mtune=i686"
  LIBDIRSUFFIX=""
```

When $ARCH="i686", both the -march and the -mtune options are set to "i686", but for $ARCH="i486", only the -march option is also set to "i486", -mtune is set to "i686", maybe that's the problem here? @petevine, can you try and edit your freeorion.SlackBuild script and set -mtune to "i486" for the i486 architecture, then try to build and report back?

**Comment by MagaTailor** [comment]
I used a binary build from the slackonly repo and won't be able to build FO myself but it's clear you've found the problem (cross-compiling or even building natively on a newer SSE(x) machine with `-march=i686` leads to this, the correct way would be to use `pentium2` or `pentiumpro` flag if it's meant for generic i686) as the current 32-bit slackware has switched to `i586`. Even though the package was labelled i586, the proper slackbuild march flag is nowhere to be found.

(the 14.1 version compiled for i486 didn't have the SIGILL issue, just the one that's been fixed here)

**Comment by Vezzra** [comment]
Ok, if I understand correctly, you're using a prebuilt binary that has (apparently mistakenly) been built for i686 and later architectures, which caused the SIGILL when you tried to run it on your system.

So, whoever produced these prebuilt binaries will have to adjust the build script accordingly, or the maintainer of the slackbuild package has to do that for the build script hosted on slackbuilds.org (so anyone who uses it to produce prebuilt binaries will get ones that won't cause SIGILL crashes on i586 architectures).

@petevine, can you report those findings back to whoever needs to know to make the necessary adjustments? Anything else you need from me or I can assist you with?

**Comment by MagaTailor** [comment]
Thanks, that's exactly what's needed - the slackbuild author has redirected me here so I'm sure he's reading the solution as well. The script must have defaulted to i686 after encountering an unknown arch (i586 in the current slackware)

**Comment by Vezzra** [comment]
> the slackbuild author has redirected me here so I'm sure he's reading the solution as well

Well, I'm not entirely sure if he actively monitors the discussion here, it's a closed issue after all.

@dslackw, can you confirm that you've taken notice of this issue and our discussion about it here and implemented the fix? Just so I can remove this from my todo list... :wink:

**Comment by ghost** [comment]
Slackware -current not supported from SBo repository.
http://slackbuilds.org/faq/
FAQ 15

**Comment by MagaTailor** [comment]
I've notified the author.

We're talking about this repo, as you can see it has those two versions (14.1 and current) covered:

http://packages.slackonly.com/pub/packages/

The fix is simple, just add a new elif section for `-march=i586`

```
elif [ "$ARCH" = "i586" ]; then
  SLKCFLAGS="-O2 -march=i586 -mtune=i686"
  LIBDIRSUFFIX=""
```

**Comment by Vezzra** [comment]
Ok, I see. So, if I understand correctly, the prebuilt binary for 14.1 has been working anyway, it was only the one for -current that was messed up (the build script not adjusted correctly). Well, I guess you guys have to take it from here.

Let me know if you need any further assistance from me (as far as I'm able to provide it, that is).

**Comment by ghost** [comment]
The script has been made to work correctly in version Slackware -stable 14.1, as the repository ( slackbuilds.org ) requires. Unfortunately I can not support the version -current because there are major changes in the distribution structure. If Slackware -current users want to run this script you need to make changes themselves. I think this issue closed.
